main/distributed-programming.md

```diff
@@ -50,37 +58,60 @@ lockdown(); // turns the current realm into an immutable realm
+realm with compartments providing just enough authority to create
+useful and secure contracts (but not enough authority to do anything
+unintended or harmful to the participants of the smart contract). 
+This means you cannot send email, write to disk, or visit arbitrary web
+pages, etc. from code inside a contract.
 
+## `harden()`
 
-### harden
-
-`harden` is a function provided by the
-[`@agoric/harden` package](https://www.npmjs.com/package/@agoric/harden).  It ensures
-that external callers of hardened objects can only interact with them through
-functions present in their API.  `harden` does a transitive version of
+`harden()` ensures that external callers of a hardened object can only
+interact with it through functions in the object’s API.
+ `harden()` is an enhanced transitive version of 
 [`Object.freeze`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/freeze),
 which only locks up an object's own properties.
 
-All objects that will be transferred across a trust boundary must have their API
-surface frozen (usually by calling harden). This ensures that other objects can only
-interact with them through their defined method interface. CapTP, our communications
-layer for passing references to distributed objects, enforces this at vat boundaries.
+All objects that will be transferred across a trust boundary must have
+their API surface frozen (usually by calling `harden()`). This ensures
+other objects can only interact with them through their defined
+method interface. CapTP, our communications layer for passing
+references to distributed objects, enforces this at vat boundaries.
+
+The general rule is that if you make a new object and give it to someone else (and don't immediately forget it yourself),
+you should give them `harden(obj)` instead the raw object.
+This prevents someone from adding/deleting the properties or prototypes of that object.
+It doesn't make them immutable (`harden(new Map()) ` still behaves like a normal mutable `Map`), but it means
+their methods stay the same and can't be surprisingly changed by someone else
+
+Defined objects (`mint`, `issuer`, `zcf,` etc.) shouldn't need hardening as their constructors should do that work. It's mainly records, callbacks, and ephemeral objects that need hardening.
+
+You can send a message to a hardened object. If it's a record, you can extract values from it.
 
-After you've [installed](https://docs.npmjs.com/cli/install) the [`@agoric/harden` package](https://www.npmjs.com/package/@agoric/harden), you can use it this way:
+You have to harden a class before you harden any of its instances; i.e. it takes two separate steps
+to harden both a class and its instances. `harden()` does transitive freezing by following the object’s
+own properties (as opposed to properties it inherited), and the objects whose own properties refer to them,
+and so forth. Harden a base class before hardening classes that inherit from it.
```

Comment:
I think we talk about classes and then how the freezing works, and then classes again. Can we rearrange?